File: 0260-single-number-iii/0260-single-number-iii.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var singleNumber = function(nums) {
    var x = 0;
    for(let i = 0; i < nums.length; i++)
        x ^= nums[i];
    var aX = x, p = 0;
    while(true)
    {
        if(aX % 2 != 0)
            break;
        p += 1;
        aX /= 2;
    }
    var n1 = 0, n2 = 0;
    for(var j = 0; j < nums.length; j++)
    {
        if((nums[j] & Math.pow(2, p)) == 0)
            n1 = n1 ^ nums[j];
        else
            n2 = n2 ^ nums[j];
    }
    return [n1, n2];
};

File: 0260-single-number-iii/0260-single-number-iii.rs
impl Solution {
    pub fn single_number(nums: Vec<i32>) -> Vec<i32> {
        let mut x : i32 = 0;
        for i in 0..nums.len()
        {
            x ^= nums[i];
        }
        let mut aX : i32 = x;
        let mut p : i32 = 1;
        loop
        {
            if aX % 2 != 0
            {
                break;
            }
            p <<= 1;
            aX /= 2;
        }
        let mut n1 : i32 = 0;
        let mut n2 : i32 = 0;
        for i in 0..nums.len()
        {
            if(nums[i] & p == 0)
            {
                n1 ^= nums[i];
            }
            else
            {
                n2 ^= nums[i];
            }
        }
        return [n1, n2].to_vec();
    }
}